src/metabase/query_processor/middleware/upgrade_field_literals.clj

```diff
@@ -1,36 +1,113 @@
+        ;; a developer-facing warning. Things will still work and this should be fixed on the FE, but we don't need to
+        ;; blow up prod logs
+        [:field (id :guard integer?) new-options]
+        (u/prog1 [:field id (merge new-options (dissoc options :base-type))]
+          (when (and (not config/is-prod?)
+                     (not (has-a-native-source-query-at-some-level? inner-query)))
+            ;; don't i18n this because it's developer-facing only.
+            (warn-once
+             (str "Warning: query is using a [:field <string> ...] clause to refer to a Field in an MBQL source query."
+                  \newline
+                  "Use [:field <integer> ...] clauses to refer to Fields in MBQL source queries."
+                  \newline
+                  "We will attempt to fix this, but it may lead to incorrect queries."
+                  \newline
+                  "See https://github.com/metabase/metabase/issues/16389#issuecomment-1013780973 for more information."
+                  \newline
+                  (str "Clause:       " (pr-str field-clause))
+                  \newline
+                  (str "Corrected to: " (pr-str <>))))))
+
+        ;; Otherwise the Field clause in the source query uses a string Field name as well, but that name differs from
+        ;; the one in `source-aliases`. Will this work? Not sure whether or not we need to log something about this.
+        [:field (field-name :guard string?) new-options]
+        (u/prog1 [:field field-name (merge new-options (dissoc options :base-type))]
+          (warn-once
+           (trs "Warning: clause {0} does not match a column in the source query. Attempting to correct this to {1}"
+                (pr-str field-clause)
+                (pr-str <>)))))
+
+      ;; If the field name exists in the ACTUAL names returned by the source query then we're g2g and don't need to
+      ;; complain about anything.
+      (contains? source-aliases field-name)
+      field-clause
+
+      ;; no matching Field ref means there's no column with this name in the source query. The query may not work, so
+      ;; log a warning about it. This query is probably not going to work so we should let everyone know why.
+      :else
+      (do
+        (warn-once
+         (trs "Warning: clause {0} refers to a Field that may not be present in the source query. Query may not work as expected. Found: {1}"
+              (pr-str field-clause) (pr-str (or (not-empty source-aliases)
+                                                (set (keys field-name->field))))))
+        field-clause))))
 
 (defn- upgrade-field-literals-one-level [{:keys [source-metadata], :as inner-query}]
-  (let [field-name->field (u/key-by :name source-metadata)]
-    ;; look for `field` clauses that use a string name...
+  (let [source-aliases    (into #{} (comp (map :source_alias) (filter some?)) source-metadata)
```

Comment:
`(comp (map :source_alias) (filter some?))` is just `(keep :source_alias)`

Reply:
```suggestion
  (let [source-aliases    (into #{} (keep :source_alias) source-metadata)
```